packages/react-utilities/src/compose/resolveShorthand.ts

```diff
@@ -1,36 +1,40 @@
 import { isValidElement } from 'react';
-import type { UnknownSlotProps, ShorthandProps } from './types';
+import type { UnknownSlotProps, SlotShorthandValue } from './types';
 
-// eslint-disable-next-line @typescript-eslint/no-explicit-any
-export type ResolveShorthandOptions<Props extends Record<string, any>, Required extends boolean = false> = {
+export type ResolveShorthandOptions<Props, Required extends boolean = false> = {
   required?: Required;
   defaultProps?: Props;
 };
 
+export type ResolveShorthandFunction<Props extends UnknownSlotProps = UnknownSlotProps> = {
+  <P extends Props | null>(value: P | SlotShorthandValue | undefined, options?: ResolveShorthandOptions<P, true>): P;
+  <P extends Props | null>(value: P | SlotShorthandValue | undefined, options?: ResolveShorthandOptions<P, boolean>):
+    | P
+    | undefined;
+};
+
 /**
  * Resolves shorthands into slot props, to ensure normalization of the signature
  * being passed down to getSlots method
  * @param value - the base shorthand props
  * @param options - options to resolve shorthand props
  */
-export function resolveShorthand<Props extends UnknownSlotProps, Required extends boolean = false>(
-  value: ShorthandProps<Props>,
-  options?: ResolveShorthandOptions<Props, Required>,
-): Required extends false ? Props | undefined : Props {
+export const resolveShorthand: ResolveShorthandFunction = (value, options) => {
   const { required = false, defaultProps } = options || {};
-  if (value === null || (value === undefined && !required)) {
-    return undefined as Required extends false ? Props | undefined : never;
+  if (value === null) {
+    return null;
+  }
```

Comment:
I have already mentioned this, but what is the reason to return `null`? Do we really use this somehow?

Can we simply return `undefined` to reduce amount of returned types?

```suggestion
  if (value === null) {
    return undefined;
  }
```

Reply:
The return type of this function is `Props` (if required: true), or `Props | undefined` otherwise. The `null` in the return type comes from the `Props` type if and only if Props includes null (which is the core change in this PR).

Returning `undefined` in the case of `null` props would require some type gymnastics to replace `null` with `undefined` in the Props type (both here and in ComponentState). 

We also have `required: true`, which guarantees that this won't return `undefined`, but specifically does not guarantee it won't return `null`, by design. A "required" slot can still be "nullable".

In summary: I think keeping them separate is worthwhile. And since both null and undefined are falsey, it shouldn't affect most places that just do e.g. `if (state.slot) { ... }`

Reply:
> And since both null and undefined are falsey, it shouldn't affect most places that just do e.g. if (state.slot) { ... }

The problem is that `null` is also an object:

```js
const state = { iconA: null, iconB: undefined, iconC: {} }

typeof state.iconA // 'object'
typeof state.iconB // 'undefined'
typeof state.iconC // 'object'
```

IMO it's better to reduce the shape as from consumer perspective I don't see any reason to have as result of `resolveShorthand()`.

Reply:
That doesn't really seem like a real problem to me... But anyways, I added the `ReplaceNullWithUndefined` utility type and included it in the type of `ComponentState` and `ResolveShorthandFunction`.

Commit is here: https://github.com/microsoft/fluentui/pull/21518/commits/13a8d1df32759fa29fcf989ea1d222b6ea13edd1